File: tests/diameter_of_binary_tree_test.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def diameterOfBinaryTree(root):
    diameter = 0

    def height(node):
        nonlocal diameter
        if not node:
            return 0
        left = height(node.left)
        right = height(node.right)
        diameter = max(diameter, left + right)
        return max(left, right) + 1

    height(root)
    return diameter

def testDiameter(root, expected, testName):
    result = diameterOfBinaryTree(root)
    print(f"Test: {testName}")
    print(f"Expected: {expected}")
    print(f"Result: {result}")
    assert result == expected, f"FAILED: {testName} - Expected {expected}, got {result}"
    print("Status: PASSED\n")

def testStub():
    # This test will fail if the solution is stubbed (e.g., always returns 0)
    root = TreeNode(1, TreeNode(2), TreeNode(3))
    result = diameterOfBinaryTree(root)
    assert result != 0, "Solution appears to be missing or stubbed (returns 0 for non-trivial tree)"

def main():
    totalTests = 16  # including stub test
    passedCount = 0

    try:
        testStub()
        passedCount += 1

        # 1. Empty Tree
        testDiameter(None, 0, "Empty Tree")
        passedCount += 1

        # 2. Single Node
        t2 = TreeNode(1)
        testDiameter(t2, 0, "Single Node")
        passedCount += 1

        # 3. Left Child Only
        t3 = TreeNode(1, TreeNode(2))
        testDiameter(t3, 1, "Left Child Only")
        passedCount += 1

        # 4. Right Child Only
        t4 = TreeNode(1, None, TreeNode(2))
        testDiameter(t4, 1, "Right Child Only")
        passedCount += 1

        # 5. Balanced Tree
        t5 = TreeNode(1, TreeNode(2), TreeNode(3))
        testDiameter(t5, 2, "Balanced Tree")
        passedCount += 1

        # 6. Left Skewed
        t6 = TreeNode(1, TreeNode(2, TreeNode(3, TreeNode(4))))
        testDiameter(t6, 3, "Left Skewed")
        passedCount += 1

        # 7. Right Skewed
        t7 = TreeNode(1, None, TreeNode(2, None, TreeNode(3, None, TreeNode(4))))
        testDiameter(t7, 3, "Right Skewed")
        passedCount += 1

        # 8. Full Binary Tree
        t8 = TreeNode(1,
            TreeNode(2, TreeNode(4), TreeNode(5)),
            TreeNode(3, TreeNode(6), TreeNode(7))
        )
        testDiameter(t8, 4, "Full Binary Tree")
        passedCount += 1

        # 9. V-Shape Tree
        t9 = TreeNode(1, TreeNode(2, None, TreeNode(3)))
        testDiameter(t9, 2, "V-Shape Tree")
        passedCount += 1

        # 10. Deep Left + Right
        t10 = TreeNode(1,
            TreeNode(2, TreeNode(4, TreeNode(5))),
            TreeNode(3)
        )
        testDiameter(t10, 4, "Deep Left + Right")
        passedCount += 1

        # 11. Balanced with Subtree
        t11 = TreeNode(1,
            TreeNode(2, TreeNode(4), TreeNode(5, TreeNode(6))),
            TreeNode(3)
        )
        testDiameter(t11, 4, "Balanced with Subtree")
        passedCount += 1

        # 12. Mixed Random
        t12 = TreeNode(1,
            TreeNode(2,
                TreeNode(4, TreeNode(7)),
                TreeNode(5)
            ),
            TreeNode(3, None, TreeNode(6))
        )
        testDiameter(t12, 5, "Mixed Random")
        passedCount += 1

        # 13. Deep Ends Both
        t13 = TreeNode(1,
            TreeNode(2, TreeNode(4, TreeNode(7))),
            TreeNode(3, None, TreeNode(6, None, TreeNode(8)))
        )
        testDiameter(t13, 6, "Deep Ends Both")
        passedCount += 1

        # 14. Sparse Tree
        t14 = TreeNode(1,
            TreeNode(2, None, TreeNode(4)),
            TreeNode(3)
        )
        testDiameter(t14, 3, "Sparse Tree")
        passedCount += 1

        # 15. Complex Layout
        t15 = TreeNode(1,
            TreeNode(2,
                TreeNode(4, TreeNode(6)),
                TreeNode(5)
            ),
            TreeNode(3,
                TreeNode(7),
                TreeNode(8, TreeNode(9))
            )
        )
        testDiameter(t15, 6, "Complex Layout")
        passedCount += 1

    except AssertionError as e:
        print(str(e))
        print(f"Total Tests Passed: {passedCount} out of {totalTests}")
        exit(1)  # Fail the CI

    print(f"Total Tests Passed: {passedCount} out of {totalTests}")

if __name__ == "__main__":
    main()

File: tests/diameter_of_binary_tree_test.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

// Internal helper to calculate height
static int height(struct TreeNode* node, int* diameter) {
    if (!node) return 0;
    int left = height(node->left, diameter);
    int right = height(node->right, diameter);
    if (*diameter < left + right) *diameter = left + right;
    return (left > right ? left : right) + 1;
}

int diameterOfBinaryTree(struct TreeNode* root) {
    int diameter = 0;
    height(root, &diameter);
    return diameter;
}

struct TreeNode* newNode(int val) {
    struct TreeNode* node = malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}

bool testDiameter(struct TreeNode* root, int expected, const char* testName) {
    int result = diameterOfBinaryTree(root);
    printf("Test: %s\n", testName);
    printf("Expected: %d\n", expected);
    printf("Result: %d\n", result);
    bool passed = result == expected;
    printf("Status: %s\n\n", passed ? "PASSED" : "FAILED");
    return passed;
}

int main() {
    int passedCount = 0;
    const int totalTests = 15;

    // 1. Empty Tree
    passedCount += testDiameter(NULL, 0, "Empty Tree");

    // 2. Single Node
    struct TreeNode* t2 = newNode(1);
    passedCount += testDiameter(t2, 0, "Single Node");

    // 3. Left Child Only
    struct TreeNode* t3 = newNode(1);
    t3->left = newNode(2);
    passedCount += testDiameter(t3, 1, "Left Child Only");

    // 4. Right Child Only
    struct TreeNode* t4 = newNode(1);
    t4->right = newNode(2);
    passedCount += testDiameter(t4, 1, "Right Child Only");

    // 5. Balanced Tree
    struct TreeNode* t5 = newNode(1);
    t5->left = newNode(2);
    t5->right = newNode(3);
    passedCount += testDiameter(t5, 2, "Balanced Tree");

    // 6. Left Skewed
    struct TreeNode* t6 = newNode(1);
    t6->left = newNode(2);
    t6->left->left = newNode(3);
    t6->left->left->left = newNode(4);
    passedCount += testDiameter(t6, 3, "Left Skewed");

    // 7. Right Skewed
    struct TreeNode* t7 = newNode(1);
    t7->right = newNode(2);
    t7->right->right = newNode(3);
    t7->right->right->right = newNode(4);
    passedCount += testDiameter(t7, 3, "Right Skewed");

    // 8. Full Binary Tree
    struct TreeNode* t8 = newNode(1);
    t8->left = newNode(2);
    t8->right = newNode(3);
    t8->left->left = newNode(4);
    t8->left->right = newNode(5);
    t8->right->left = newNode(6);
    t8->right->right = newNode(7);
    passedCount += testDiameter(t8, 4, "Full Binary Tree");

    // 9. V-Shape Tree
    struct TreeNode* t9 = newNode(1);
    t9->left = newNode(2);
    t9->left->right = newNode(3);
    passedCount += testDiameter(t9, 2, "V-Shape Tree");

    // 10. Deep Left + Right
    struct TreeNode* t10 = newNode(1);
    t10->left = newNode(2);
    t10->left->left = newNode(4);
    t10->left->left->left = newNode(5);
    t10->right = newNode(3);
    passedCount += testDiameter(t10, 4, "Deep Left + Right");

    // 11. Balanced with Subtree
    struct TreeNode* t11 = newNode(1);
    t11->left = newNode(2);
    t11->left->left = newNode(4);
    t11->left->right = newNode(5);
    t11->left->right->left = newNode(6);
    t11->right = newNode(3);
    passedCount += testDiameter(t11, 4, "Balanced with Subtree");

    // 12. Mixed Random
    struct TreeNode* t12 = newNode(1);
    t12->left = newNode(2);
    t12->left->left = newNode(4);
    t12->left->left->left = newNode(7);
    t12->left->right = newNode(5);
    t12->right = newNode(3);
    t12->right->right = newNode(6);
    passedCount += testDiameter(t12, 5, "Mixed Random");

    // 13. Deep Ends Both
    struct TreeNode* t13 = newNode(1);
    t13->left = newNode(2);
    t13->left->left = newNode(4);
    t13->left->left->left = newNode(7);
    t13->right = newNode(3);
    t13->right->right = newNode(6);
    t13->right->right->right = newNode(8);
    passedCount += testDiameter(t13, 6, "Deep Ends Both");

    // 14. Sparse Tree
    struct TreeNode* t14 = newNode(1);
    t14->left = newNode(2);
    t14->left->right = newNode(4);
    t14->right = newNode(3);
    passedCount += testDiameter(t14, 3, "Sparse Tree");

    // 15. Complex Layout
    struct TreeNode* t15 = newNode(1);
    t15->left = newNode(2);
    t15->left->left = newNode(4);
    t15->left->left->left = newNode(6);
    t15->left->right = newNode(5);
    t15->right = newNode(3);
    t15->right->left = newNode(7);
    t15->right->right = newNode(8);
    t15->right->right->left = newNode(9);
    passedCount += testDiameter(t15, 6, "Complex Layout");

    printf("Total Tests Passed: %d out of %d\n", passedCount, totalTests);
    return 0;
}
